# Introduction

In this project, we will work with data from the CIA World Factbook, a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like:

- population - The population as of 2015.
- population_growth - The annual population growth rate, as a percentage.
- area - The total land and water area.

In this guided project, we will use SQL in Jupyter Notebook to explore and analyze data from this database. 

Let us first connect Jupyter Notebook to our database file:

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

# Overview of the data

Get a list of all tables in the database:

In [2]:
%%sql
SELECT *
  FROM sqlite_master
 WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

Here is an overview of the 'facts' database:

In [3]:
%%sql
SELECT *
  FROM facts
 LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Here are the descriptions for some of the columns:

- name - The name of the country.
- area - The total land and sea area of the country.
- population - The country's population.
- population_growth- The country's population growth as a percentage.
- birth_rate - The country's birth rate, or the number of births a year per 1,000 people.
- death_rate - The country's death rate, or the number of death a year per 1,000 people.
- area- The country's total area (both land and water).
- area_land - The country's land area in square kilometers.
- area_water - The country's waterarea in square kilometers.

# Summary statistics

We will start by calculating some summary statistics.

In [4]:
%%sql
SELECT MIN(population) AS min_pop,
       MAX(population) AS max_pop,
       MIN(population_growth) AS min_pop_growth,
       MAX(population_growth) max_pop_growth 
  FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

We see from the summary statistics that:

- There is a country with a population of 0.
- There is a country with a population of 7256490011 (i.e., more than 7.2 billion people).

Let us use subqueries to zoom in on just these countries without using the specific values.

# Exploring extreme values

In [5]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MIN(population)
                        FROM facts
                     );

 * sqlite:///factbook.db
Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

Antartica has a population of zero. This is a correct number.

In [6]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MAX(population)
                        FROM facts
                     );

 * sqlite:///factbook.db
Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

We see that the table contains a row for the whole world, which explains the maximum population of over 7.2 billion.

Now that we know this, we should recalculate the summary statistics we calculated earlier, while excluding the row for the whole world.

# Revisited summary statistics

In [9]:
%%sql
SELECT MIN(population) AS min_pop,
       MAX(population) AS max_pop,
       MIN(population_growth) AS min_pop_growth,
       MAX(population_growth) AS max_pop_growth 
  FROM facts
 WHERE name <> 'World'; -- <> = not equal to

 * sqlite:///factbook.db
Done.


[(0, 1367485388, 0.0, 4.02)]

We see that there is one country with a population of almost 1.4 billion people.

# Exploring density using average population and area

Let us explore the population density. Population density depends on the population and the country's area. Let us look at the average values for these two columns.

We should take care of discarding the row for the whole planet.


In [10]:
%%sql
SELECT AVG(population) AS avg_population, AVG(area) AS avg_area
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[(32242666.56846473, 555093.546184739)]

We see that the average population is around 32 million and the average area is 555 thousand square kilometers.

In [14]:
%%sql
SELECT name, (population/area) AS density
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[('Afghanistan', 49),
 ('Albania', 105),
 ('Algeria', 16),
 ('Andorra', 182),
 ('Angola', 15),
 ('Antigua and Barbuda', 209),
 ('Argentina', 15),
 ('Armenia', 102),
 ('Australia', 2),
 ('Austria', 103),
 ('Azerbaijan', 112),
 ('Bahamas, The', 23),
 ('Bahrain', 1771),
 ('Bangladesh', 1138),
 ('Barbados', 675),
 ('Belarus', 46),
 ('Belgium', 370),
 ('Belize', 15),
 ('Benin', 92),
 ('Bhutan', 19),
 ('Bolivia', 9),
 ('Bosnia and Herzegovina', 75),
 ('Botswana', 3),
 ('Brazil', 23),
 ('Brunei', 74),
 ('Bulgaria', 64),
 ('Burkina Faso', 69),
 ('Burma', 83),
 ('Burundi', 385),
 ('Cambodia', 86),
 ('Cameroon', 49),
 ('Canada', 3),
 ('Cabo Verde', 135),
 ('Central African Republic', 8),
 ('Chad', None),
 ('Chile', 23),
 ('China', 142),
 ('Colombia', 41),
 ('Comoros', 349),
 ('Congo, Democratic Republic of the', 33),
 ('Congo, Republic of the', 13),
 ('Costa Rica', 94),
 ("Cote d'Ivoire", 72),
 ('Croatia', 78),
 ('Cuba', 99),
 ('Cyprus', 128),
 ('Czech Republic', 134),
 ('Denmark', 129),
 ('Djibouti', 35),
 ('Dominica', 98),
 ('Dominican Republic', 215),
 ('Ecuador', 55),
 ('Egypt', 88),
 ('El Salvador', 291),
 ('Equatorial Guinea', 26),
 ('Eritrea', 55),
 ('Estonia', 27),
 ('Ethiopia', 90),
 ('Fiji', 49),
 ('Finland', 16),
 ('France', 103),
 ('Gabon', 6),
 ('Gambia, The', 174),
 ('Georgia', 70),
 ('Germany', 226),
 ('Ghana', 110),
 ('Greece', 81),
 ('Grenada', 321),
 ('Guatemala', 137),
 ('Guinea', 47),
 ('Guinea-Bissau', 47),
 ('Guyana', 3),
 ('Haiti', 364),
 ('Honduras', 78),
 ('Hungary', 106),
 ('Iceland', 3),
 ('India', 380),
 ('Indonesia', 134),
 ('Iran', 49),
 ('Iraq', 84),
 ('Ireland', 69),
 ('Israel', 387),
 ('Italy', 205),
 ('Jamaica', 268),
 ('Japan', 335),
 ('Jordan', 90),
 ('Kazakhstan', 6),
 ('Kenya', 79),
 ('Kiribati', 130),
 ('Korea, North', 207),
 ('Korea, South', 492),
 ('Kosovo', 171),
 ('Kuwait', 156),
 ('Kyrgyzstan', 28),
 ('Laos', 29),
 ('Latvia', 30),
 ('Lebanon', 594),
 ('Lesotho', 64),
 ('Liberia', 37),
 ('Libya', 3),
 ('Liechtenstein', 235),
 ('Lithuania', 44),
 ('Luxembourg', 220),
 ('Macedonia', 81),
 ('Madagascar', 40),
 ('Malawi', 151),
 ('Malaysia', 92),
 ('Maldives', 1319),
 ('Mali', 13),
 ('Malta', 1310),
 ('Marshall Islands', 398),
 ('Mauritania', 3),
 ('Mauritius', 656),
 ('Mexico', 61),
 ('Micronesia, Federated States of', 149),
 ('Moldova', 104),
 ('Monaco', 15267),
 ('Mongolia', 1),
 ('Montenegro', 46),
 ('Morocco', 74),
 ('Mozambique', 31),
 ('Namibia', 2),
 ('Nauru', 454),
 ('Nepal', 214),
 ('Netherlands', 407),
 ('New Zealand', 16),
 ('Nicaragua', 45),
 ('Niger', None),
 ('Nigeria', 196),
 ('Norway', 16),
 ('Oman', 10),
 ('Pakistan', 250),
 ('Palau', 46),
 ('Panama', 48),
 ('Papua New Guinea', 14),
 ('Paraguay', 16),
 ('Peru', 23),
 ('Philippines', 336),
 ('Poland', 123),
 ('Portugal', 117),
 ('Qatar', 189),
 ('Romania', 90),
 ('Russia', 8),
 ('Rwanda', 480),
 ('Saint Kitts and Nevis', 198),
 ('Saint Lucia', 266),
 ('Saint Vincent and the Grenadines', 263),
 ('Samoa', 69),
 ('San Marino', 541),
 ('Sao Tome and Principe', 201),
 ('Saudi Arabia', 12),
 ('Senegal', 71),
 ('Serbia', 92),
 ('Seychelles', 203),
 ('Sierra Leone', 81),
 ('Singapore', 8141),
 ('Slovakia', 111),
 ('Slovenia', 97),
 ('Solomon Islands', 21),
 ('Somalia', 16),
 ('South Africa', 44),
 ('South Sudan', 18),
 ('Spain', 95),
 ('Sri Lanka', 336),
 ('Sudan', 19),
 ('Suriname', 3),
 ('Swaziland', 82),
 ('Sweden', 21),
 ('Switzerland', 196),
 ('Syria', 92),
 ('Tajikistan', 56),
 ('Tanzania', 53),
 ('Thailand', 132),
 ('Timor-Leste', 82),
 ('Togo', 132),
 ('Tonga', 142),
 ('Trinidad and Tobago', 238),
 ('Tunisia', 67),
 ('Turkey', 101),
 ('Turkmenistan', 10),
 ('Tuvalu', 418),
 ('Uganda', 153),
 ('Ukraine', 73),
 ('United Arab Emirates', 69),
 ('United Kingdom', 263),
 ('United States', 32),
 ('Uruguay', 18),
 ('Uzbekistan', 65),
 ('Vanuatu', 22),
 ('Holy See (Vatican City)', None),
 ('Venezuela', 32),
 ('Vietnam', 284),
 ('Yemen', 50),
 ('Zambia', 20),
 ('Zimbabwe', 36),
 ('Taiwan', 650),
 ('European Union', 118),
 ('Ashmore and Carti

In [25]:
%%sql
SELECT 
    MIN(population/area) AS min_density, 
    MAX(population/area) AS max_density, 
    AVG(population/area) AS mean_density
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[(0, 21168, 419.14830508474574)]

To finish, we will identify countries that have:

- Above average values for population.
- Below average values for area.

In [26]:
%%sql
SELECT *
  FROM facts
 WHERE population > (SELECT AVG(population)
                       FROM facts
                    )
   AND area < (SELECT AVG(area)
                 FROM facts
);

 * sqlite:///factbook.db
Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

Some of these countries are generally known to be densely populated, so we have confidence in our results.